File: code_snippets/java/src/main/java/get_started/Tour.java
package get_started;

import dev.restate.sdk.Context;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.Service;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.http.vertx.RestateHttpServer;
import java.time.Duration;

@Service
public class Tour {

  @Handler
  public void myFn(Context ctx) {
    String ticketId = "123";

    // <start_sleep>
    ctx.sleep(Duration.ofMinutes(15));
    // <end_sleep>

    // <start_sleep_and_send>
    ctx.sleep(Duration.ofMinutes(15));
    TicketObjectClient.fromContext(ctx, ticketId).send().unreserve();
    // <end_sleep_and_send>
  }

  // <start_uuid>
  @Handler
  public boolean handle(Context ctx, CheckoutRequest request) {
    // !mark
    String idempotencyKey = ctx.random().nextUUID().toString();
    // !mark
    System.out.println("My idempotency key: " + idempotencyKey);

    // !mark
    throw new IllegalStateException("The handler failed");
  }

  // <end_uuid>

  public static void main(String[] args) {
    RestateHttpServer.listen(Endpoint.bind(new Tour()).bind(new TicketObject()).build());
  }
}

@Service
class CheckoutService {
  // <start_checkout>
  @Handler
  public boolean handle(Context ctx, CheckoutRequest request) {
    // !mark
    double totalPrice = request.getTickets().size() * 40.0;

    String idempotencyKey = ctx.random().nextUUID().toString();

    // !mark
    boolean success =
        // !mark
        ctx.run(
            // !mark
            Boolean.class,
            // !mark
            () -> PaymentClient.get().call(idempotencyKey, totalPrice));

    return success;
  }
  // <end_checkout>

}

class PaymentClient {

  public static PaymentClient get() {
    return new PaymentClient();
  }

  public boolean call(String idempotencyKey, double totalPrice) {
    return true;
  }
}

File: code_snippets/python/src/get_started/tour.py
# Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
#
# This file is part of the Restate examples,
# which is released under the MIT license.
#
# You can find a copy of the license in the file LICENSE
# in the root directory of this repository or package or at
# https://github.com/restatedev/examples/
import uuid
from datetime import timedelta
from typing import TypedDict, List

from restate import VirtualObject
from restate.context import ObjectContext, Serde
from restate.service import Service

cart = VirtualObject("CartObject")


@cart.handler()
async def my_handler(ctx: ObjectContext, order: str) -> bool:
    ticket_id = ""

    # <start_sleep>
    await ctx.sleep(delta=timedelta(minutes=15))
    # <end_sleep>

    # <start_sleep_and_send>
    await ctx.sleep(delta=timedelta(minutes=15))
    ctx.object_send(unreserve, key=ticket_id, arg=None)
    # <end_sleep_and_send>

    return True


ticket = VirtualObject("TicketObject")


@ticket.handler()
async def unreserve(ctx: ObjectContext) -> bool:
    return True


class Order(TypedDict):
    user_id: str
    tickets: List[str]


checkout = Service("CheckoutService")


# <start_uuid>
@checkout.handler()
async def handle(ctx: ObjectContext, order: Order) -> bool:
    # !mark(1:3)
    idempotency_key = await ctx.run("idempotency_key", lambda: str(uuid.uuid4()))
    print("My idempotency key is: ", idempotency_key)
    raise Exception("Something happened!")

    return True


# <end_uuid>
